File: apps/vinto/src/client/bot/vinto-round-solver.ts
// services/vinto-round-solver.ts

import { Card, CardAction, getCardAction, getCardValue, Pile } from '@/shared';
import { BotMemory } from './bot-memory';

/**
 * Result of Vinto validation analysis
 */
export interface VintoValidationResult {
  shouldCallVinto: boolean;
  callerScore: number;
  worstCaseOpponentScore: number;
  confidence: number; // 0-1, how confident we are in the analysis
  reason: string;
}

/**
 * Represents a player's state for Vinto analysis
 */
interface PlayerAnalysisState {
  id: string;
  knownCards: Card[];
  unknownCardCount: number;
  hasActionCards: boolean;
  actionCardTypes: CardAction[];
}

/**
 * VintoRoundSolver - Performs worst-case analysis for Vinto calls
 *
 * This solver checks if calling Vinto is safe by simulating the best possible
 * moves opponents can make with their known action cards and hidden cards.
 *
 * The goal: Ensure that even if opponents play optimally to minimize their scores,
 * the bot's score is still lower than the best opponent's score.
 */
export class VintoRoundSolver {
  constructor(_botId: string, private botMemory: BotMemory) {}

  /**
   * Validate if calling Vinto is safe
   * Returns detailed analysis of whether the bot should call Vinto
   */
  validateVintoCall(
    botCards: Card[],
    opponents: Array<{
      id: string;
      cardCount: number;
    }>,
    discardPile: Pile
  ): VintoValidationResult {
    // Step 1: Calculate bot's actual score
    const callerScore = this.calculateScore(botCards);

    // Step 2: Build analysis state for each opponent
    const opponentStates = opponents.map((opp) =>
      this.buildOpponentState(opp.id, opp.cardCount)
    );

    // Step 3: Calculate worst-case scenario for each opponent
    // (best score they can achieve through optimal play)
    const opponentWorstCases = opponentStates.map((oppState) =>
      this.calculateWorstCaseScore(oppState, discardPile)
    );

    // Step 4: Find the best opponent (lowest possible score)
    const worstCaseOpponentScore = Math.min(...opponentWorstCases);

    // Step 5: Determine if Vinto call is safe
    // Bot wins if: botScore < worstCaseOpponentScore
    const shouldCallVinto = callerScore < worstCaseOpponentScore;

    // Step 6: Calculate confidence based on information quality
    const confidence = this.calculateConfidence(opponentStates);

    // Step 7: Generate reason for decision
    const reason = this.generateReason(
      shouldCallVinto,
      callerScore,
      worstCaseOpponentScore,
      opponentStates,
      confidence
    );

    return {
      shouldCallVinto,
      callerScore,
      worstCaseOpponentScore,
      confidence,
      reason,
    };
  }

  /**
   * Build analysis state for an opponent
   */
  private buildOpponentState(
    opponentId: string,
    cardCount: number
  ): PlayerAnalysisState {
    const knownCards: Card[] = [];
    const actionCardTypes: CardAction[] = [];
    let unknownCardCount = 0;

    const playerMemory = this.botMemory.getPlayerMemory(opponentId);

    for (let pos = 0; pos < cardCount; pos++) {
      const memory = playerMemory.get(pos);
      if (memory && memory.confidence > 0.5 && memory.card) {
        knownCards.push(memory.card);

        // Track action cards - get action from card rank config
        const action = getCardAction(memory.card.rank);
        if (action) {
          actionCardTypes.push(action);
        }
      } else {
        unknownCardCount++;
      }
    }

    return {
      id: opponentId,
      knownCards,
      unknownCardCount,
      hasActionCards: actionCardTypes.length > 0,
      actionCardTypes,
    };
  }

  /**
   * Calculate worst-case (best for opponent) score they can achieve
   *
   * This assumes opponents play optimally:
   * 1. Use action cards to peek/swap to minimize their score
   * 2. Unknown cards are assumed to be low-value cards (pessimistic for bot)
   * 3. Opponents will use swaps to exchange high cards with low cards
   */
  private calculateWorstCaseScore(
    opponent: PlayerAnalysisState,
    discardPile: Pile
  ): number {
    // Start with known cards score
    const knownScore = opponent.knownCards.reduce(
      (sum, card) => sum + card.value,
      0
    );

    // Calculate best possible value for unknown cards
    const bestPossibleUnknownValue =
      this.calculateBestPossibleUnknownValue(discardPile);

    // Assume unknown cards are the best possible (worst case for bot)
    const unknownScore = opponent.unknownCardCount * bestPossibleUnknownValue;

    // Apply action card benefits (opponent plays optimally)
    let actionBenefit = 0;

    // Peek actions: Allow opponent to gain perfect information
    // (doesn't directly reduce score, but enables better swaps)

    // Swap actions: Each swap can replace a high card with a low card
    const swapActions = opponent.actionCardTypes.filter(
      (action) => action === 'swap-cards' || action === 'peek-and-swap'
    ).length;

    // Conservative estimate: each swap action can reduce score by ~5 points
    // (swapping a 10 for a 2, or a 9 for a Joker, etc.)
    const swapBenefit = swapActions * 5;
    actionBenefit += swapBenefit;

    // King (declare-action): Can be used as any action card
    const kingActions = opponent.actionCardTypes.filter(
      (action) => action === 'declare-action'
    ).length;
    if (kingActions > 0) {
      // Assume King is used as a swap action for maximum benefit
      actionBenefit += kingActions * 5;
    }

    // Force-draw: Doesn't help opponent's score, skip

    // Calculate final worst-case score
    const worstCaseScore = Math.max(
      0,
      knownScore + unknownScore - actionBenefit
    );

    return worstCaseScore;
  }

  /**
   * Calculate the best possible value for unknown cards
   * (lowest possible, which is worst for bot calling Vinto)
   */
  private calculateBestPossibleUnknownValue(_discardPile: Pile): number {
    const distribution = this.botMemory.getCardDistribution();

    // Build list of remaining card values, weighted by count
    const remainingValues: number[] = [];
    for (const [rank, count] of distribution) {
      if (count > 0) {
        const value = getCardValue(rank);
        for (let i = 0; i < count; i++) {
          remainingValues.push(value);
        }
      }
    }

    if (remainingValues.length === 0) {
      // No information, assume neutral value
      return 6;
    }

    // Sort values (ascending) to find best possible cards
    remainingValues.sort((a, b) => a - b);

    // Calculate average of bottom 30% (pessimistic assumption)
    // Opponent likely has low-value cards in unknown positions
    const bottomPercentile = Math.ceil(remainingValues.length * 0.3);
    const bestValues = remainingValues.slice(0, Math.max(1, bottomPercentile));

    const avgBestValue =
      bestValues.reduce((sum, v) => sum + v, 0) / bestValues.length;

    return avgBestValue;
  }

  /**
   * Calculate confidence in the analysis
   * Based on how much information we have about opponents
   */
  private calculateConfidence(opponents: PlayerAnalysisState[]): number {
    let totalCards = 0;
    let knownCards = 0;

    for (const opp of opponents) {
      totalCards += opp.knownCards.length + opp.unknownCardCount;
      knownCards += opp.knownCards.length;
    }

    if (totalCards === 0) {
      return 0.3; // Low confidence with no information
    }

    // Confidence = percentage of opponent cards known
    const knowledgeRatio = knownCards / totalCards;

    // Scale: 30% base confidence, up to 95% with perfect knowledge
    return 0.3 + knowledgeRatio * 0.65;
  }

  /**
   * Calculate total score from cards
   */
  private calculateScore(cards: Card[]): number {
    return cards.reduce((sum, card) => sum + card.value, 0);
  }

  /**
   * Generate human-readable reason for the decision
   */
  private generateReason(
    shouldCall: boolean,
    callerScore: number,
    worstCaseScore: number,
    opponents: PlayerAnalysisState[],
    confidence: number
  ): string {
    const margin = Math.abs(callerScore - worstCaseScore);
    const knownOpponentCards = opponents.reduce(
      (sum, opp) => sum + opp.knownCards.length,
      0
    );
    const totalOpponentCards = opponents.reduce(
      (sum, opp) => sum + opp.knownCards.length + opp.unknownCardCount,
      0
    );

    if (shouldCall) {
      return `Safe to call Vinto. Score ${callerScore} < ${worstCaseScore} (margin: ${margin}). Confidence: ${(
        confidence * 100
      ).toFixed(
        0
      )}% (${knownOpponentCards}/${totalOpponentCards} opponent cards known).`;
    } else {
      const actionCardsCount = opponents.reduce(
        (sum, opp) => sum + opp.actionCardTypes.length,
        0
      );
      return `Risky to call Vinto. Score ${callerScore} >= ${worstCaseScore} (deficit: ${margin}). Opponents have ${actionCardsCount} known action cards. Confidence: ${(
        confidence * 100
      ).toFixed(0)}%.`;
    }
  }
}

File: kmp/shared/bot/src/commonMain/kotlin/game/vinto/bot/VintoRoundSolver.kt
package game.vinto.bot

import game.vinto.shapes.Card
import game.vinto.shapes.CardAction
import game.vinto.shapes.getCardAction
import game.vinto.shapes.getCardValue
import kotlin.math.ceil
import kotlin.math.max
import kotlin.math.min
import kotlin.math.round

/**
 * Whether calling Vinto is actually safe, ported from
 * `packages/bot/src/lib/vinto-round-solver.ts`.
 *
 * Calling Vinto is the one decision the bot cannot take back. Every other mistake costs a
 * few points; this one hands the round to whoever is quietly sitting on a better hand, and
 * the coalition still gets a turn afterwards to improve it. So the solver does not ask
 * whether the bot is *probably* ahead — it asks whether it is still ahead when every
 * opponent gets everything they could want:
 *
 * - unseen cards are assumed to be among the best still in the deck, not average ones;
 * - every swap action an opponent is known to hold is assumed to land well;
 * - a King is assumed to be spent as a swap, which is its most valuable use here.
 *
 * The answer comes with a [VintoValidationResult.confidence], because "safe against the
 * worst case" means much less when the bot has seen almost none of the table.
 */
class VintoRoundSolver(private val botMemory: BotMemory) {

    /** A memory below this is a hunch; the solver counts the card as unseen instead. */
    private companion object {

        /** What one swap action is assumed to save its owner — a 10 traded for a 2, say. */
        const val SWAP_BENEFIT = 5.0

        /** Used when the bot has lost track of the deck entirely. */
        const val NEUTRAL_CARD_VALUE = 6.0

        /** Unseen cards are assumed to come from the best of what is left, not the middle. */
        const val OPTIMISTIC_PERCENTILE = 0.3

        /** Knowing nothing is not the same as knowing the answer is no. */
        const val NO_INFORMATION_CONFIDENCE = 0.3
        const val BASE_CONFIDENCE = 0.3
        const val CONFIDENCE_FROM_KNOWLEDGE = 0.65

        const val PERCENT = 100
    }

    data class VintoValidationResult(
        val shouldCallVinto: Boolean,
        val callerScore: Int,
        val worstCaseOpponentScore: Double,
        /** 0-1: how much of the table the analysis is actually based on. */
        val confidence: Double,
        val reason: String,
    )

    /** An opponent as the analysis sees them: what is known, what is not, and what they hold. */
    private data class PlayerAnalysisState(
        val id: String,
        val knownCards: List<Card>,
        val unknownCardCount: Int,
        val actionCardTypes: List<CardAction>,
    ) {
        val totalCards: Int get() = knownCards.size + unknownCardCount
    }

    data class OpponentHand(val id: String, val cardCount: Int)

    fun validateVintoCall(botCards: List<Card>, opponents: List<OpponentHand>): VintoValidationResult {
        val callerScore = botCards.sumOf { it.value }
        val opponentStates = opponents.map { buildOpponentState(it.id, it.cardCount) }

        // The bot is only safe against the *best* opponent, so the worst case is the minimum.
        val worstCaseOpponentScore = opponentStates
            .map { calculateWorstCaseScore(it) }
            .minOrNull()
            ?: Double.MAX_VALUE

        val shouldCallVinto = callerScore < worstCaseOpponentScore
        val confidence = calculateConfidence(opponentStates)

        return VintoValidationResult(
            shouldCallVinto = shouldCallVinto,
            callerScore = callerScore,
            worstCaseOpponentScore = worstCaseOpponentScore,
            confidence = confidence,
            reason = generateReason(
                shouldCall = shouldCallVinto,
                callerScore = callerScore,
                worstCaseScore = worstCaseOpponentScore,
                opponents = opponentStates,
                confidence = confidence,
            ),
        )
    }

    private fun buildOpponentState(opponentId: String, cardCount: Int): PlayerAnalysisState {
        val playerMemory = botMemory.getPlayerMemory(opponentId)

        val knownCards = mutableListOf<Card>()
        val actionCardTypes = mutableListOf<CardAction>()
        var unknownCardCount = 0

        for (position in 0 until cardCount) {
            val memory = playerMemory[position]
            if (memory != null && memory.confidence > TRUSTED_CONFIDENCE) {
                knownCards += memory.card
                getCardAction(memory.card.rank)?.let { actionCardTypes += it }
            } else {
                unknownCardCount++
            }
        }

        return PlayerAnalysisState(opponentId, knownCards, unknownCardCount, actionCardTypes)
    }

    /**
     * The best score this opponent could still reach, which is the worst news for the caller.
     *
     * Peeks are not counted directly — they reveal rather than reduce — but they are what
     * makes the swap assumption reasonable, since an opponent who can see their hand knows
     * which card to trade away.
     */
    private fun calculateWorstCaseScore(opponent: PlayerAnalysisState): Double {
        val knownScore = opponent.knownCards.sumOf { it.value }.toDouble()
        val unknownScore = opponent.unknownCardCount * calculateBestPossibleUnknownValue()

        val swaps = opponent.actionCardTypes.count {
            it == CardAction.SWAP_CARDS || it == CardAction.PEEK_AND_SWAP
        }
        // A King declares any rank, so it is assumed to be spent on the best of them.
        val kings = opponent.actionCardTypes.count { it == CardAction.DECLARE_ACTION }
        val actionBenefit = (swaps + kings) * SWAP_BENEFIT

        // Action benefit is an assumption, so it may improve a hand only to zero, never
        // *below* it — two Kings are not worth -10. But a hand *observed* to be negative
        // (a Joker seen at the table) really is negative, and pretending otherwise would
        // let a caller on zero walk into a hand it has already seen beating it.
        val observedScore = knownScore + unknownScore
        return max(min(0.0, observedScore), observedScore - actionBenefit)
    }

    /**
     * What an unseen card is assumed to be worth: the average of the best 30% still out
     * there, rather than the average of everything.
     *
     * This is the asymmetry the whole solver rests on. Assuming an opponent's blind cards are
     * average makes calling Vinto look safe far too often, because the hands that beat you
     * are exactly the ones that are not average.
     */
    private fun calculateBestPossibleUnknownValue(): Double {
        val remainingValues = botMemory.getCardDistribution()
            .filter { it.value > 0 }
            .flatMap { (rank, count) -> List(count) { getCardValue(rank) } }
            .sorted()

        if (remainingValues.isEmpty()) return NEUTRAL_CARD_VALUE

        val percentileSize = ceil(remainingValues.size * OPTIMISTIC_PERCENTILE).toInt()
        val best = remainingValues.take(max(1, percentileSize))

        return best.sum().toDouble() / best.size
    }

    /** How much of the table the analysis actually saw, on a 0.3-0.95 scale. */
    private fun calculateConfidence(opponents: List<PlayerAnalysisState>): Double {
        val totalCards = opponents.sumOf { it.totalCards }
        if (totalCards == 0) return NO_INFORMATION_CONFIDENCE

        val knownCards = opponents.sumOf { it.knownCards.size }
        return BASE_CONFIDENCE + (knownCards.toDouble() / totalCards) * CONFIDENCE_FROM_KNOWLEDGE
    }

    /** Diagnostic text: this ends up in logs and in the UI's explanation of a bot's call. */
    private fun generateReason(
        shouldCall: Boolean,
        callerScore: Int,
        worstCaseScore: Double,
        opponents: List<PlayerAnalysisState>,
        confidence: Double,
    ): String {
        val margin = round(kotlin.math.abs(callerScore - worstCaseScore)).toInt()
        val knownOpponentCards = opponents.sumOf { it.knownCards.size }
        val totalOpponentCards = opponents.sumOf { it.totalCards }
        val confidencePercent = round(confidence * PERCENT).toInt()
        val worstCase = round(worstCaseScore).toInt()

        return if (shouldCall) {
            "Safe to call Vinto. Score $callerScore < $worstCase (margin: $margin). " +
                "Confidence: $confidencePercent% " +
                "($knownOpponentCards/$totalOpponentCards opponent cards known)."
        } else {
            val actionCards = opponents.sumOf { it.actionCardTypes.size }
            "Risky to call Vinto. Score $callerScore >= $worstCase (deficit: $margin). " +
                "Opponents have $actionCards known action cards. Confidence: $confidencePercent%."
        }
    }
}
